### PR TITLE
Enhance feature duplication by ensuring children of relations with composition strength are duplicated too

### PR DESCRIPTION
**Pull request by nirvn**
This PR improves the feature duplication by relying on QGIS' QgsVectorLayerUtils::duplicateFeatures function which takes care of copying any child of relationships with a strength set to composition.

A little bit of custom code is added on top of this to improve on an implementation gap on the QGIS side whereas autogenerated IDs such as geopackage's fid are not functioning properly as field pairs in relations due to feature addition being in an edit buffer and not having a proper value yet.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for ef743fa0e9d28d4b72bfcf27a384d0e48272d23f: [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4494-5c3fccadb79972ad30422a9f91971c94c88c6d6b-arm64-android.apk)
<details>
<summary>Other architectures</summary>

- [arm-neon-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4494-5c3fccadb79972ad30422a9f91971c94c88c6d6b-arm-neon-android.apk)
- [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4494-5c3fccadb79972ad30422a9f91971c94c88c6d6b-x64-android.apk)
- [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4494-5c3fccadb79972ad30422a9f91971c94c88c6d6b-x86-android.apk)
- [**all access arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4494-5c3fccadb79972ad30422a9f91971c94c88c6d6b-arm64-android.apk)
- [all access arm-neon-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4494-5c3fccadb79972ad30422a9f91971c94c88c6d6b-arm-neon-android.apk)
- [all access x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4494-5c3fccadb79972ad30422a9f91971c94c88c6d6b-x64-android.apk)
- [all access x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4494-5c3fccadb79972ad30422a9f91971c94c88c6d6b-x86-android.apk)
</details>

File: src/core/utils/layerutils.cpp [hunk 1]
```diff
@@ -312,21 +312,46 @@ QgsFeature LayerUtils::duplicateFeature( QgsVectorLayer *layer, const QgsFeature
...
+
+  // we have to re-apply referenced feature attribute value to take into account value changes form the
+  // data provider (e.g. autogenerated fid)
+  const QList<QgsRelation> relations = QgsProject::instance()->relationManager()->referencedRelations( layer );
+  for ( QgsVectorLayer *chl : duplicateFeatureContextLayers )
   {
-    QgsMessageLog::logMessage( tr( "Cannot add new feature in layer \"%1\"." ).arg( layer->name() ), "QField", Qgis::Warning );
-    return QgsFeature();
+    const QgsFeatureIds fids = duplicateFeatureContext.duplicatedFeatures( chl );
+    for ( auto relation : relations )
```
Comment (m-kuhn) [inline src/core/utils/layerutils.cpp:339]:
```suggestion
    for ( const auto &relation : relations )
```

File: src/core/utils/layerutils.cpp [hunk 2]
```diff
@@ -312,21 +312,46 @@ QgsFeature LayerUtils::duplicateFeature( QgsVectorLayer *layer, const QgsFeature
...
+    {
+      if ( relation.strength() == Qgis::RelationshipStrength::Composition && relation.referencingLayer() == chl )
+      {
+        const QList<QgsRelation::FieldPair> fieldPairs = relation.fieldPairs();
+        for ( auto fieldPair : fieldPairs )
```
Comment (m-kuhn) [inline src/core/utils/layerutils.cpp:344]:
```suggestion
        for ( const auto &fieldPair : fieldPairs )
```

File: src/core/utils/layerutils.cpp [hunk 3]
```diff
@@ -312,21 +312,46 @@ QgsFeature LayerUtils::duplicateFeature( QgsVectorLayer *layer, const QgsFeature
...
+  // data provider (e.g. autogenerated fid)
+  const QList<QgsRelation> relations = QgsProject::instance()->relationManager()->referencedRelations( layer );
+  for ( QgsVectorLayer *chl : duplicateFeatureContextLayers )
   {
-    QgsMessageLog::logMessage( tr( "Cannot add new feature in layer \"%1\"." ).arg( layer->name() ), "QField", Qgis::Warning );
-    return QgsFeature();
+    const QgsFeatureIds fids = duplicateFeatureContext.duplicatedFeatures( chl );
+    for ( auto relation : relations )
+    {
+      if ( relation.strength() == Qgis::RelationshipStrength::Composition && relation.referencingLayer() == chl )
```
Comment (m-kuhn) [inline src/core/utils/layerutils.cpp:341]: I don't understand this completely, why does the relationship strength matter here? Can't we just update all fids from the context without checking any relationship strength here?

Comment (nirvn) [reply src/core/utils/layerutils.cpp:341]: True

**Comment by nirvn** [comment]
@m-kuhn , thanks for the review, comments addressed.

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**